### PR TITLE
Add rule for nested workspaces

### PR DESCRIPTION
**Pull request by johnwiseheart**
Ensure that the `workspaces` field in the root `package.json` is correct by checking all of the packages in the repo, and ensuring that the workspaces in the `workspaces` field cover all the found packages.

Not sure what the deal is with the yarn.lock file - seems like things in there are mostly `yarnpkg` and `palantir` registries but I can't get my registry to set to the yarnpkg one (and clearly the palantir one is wrong).

This rule doesn't have a fixer. I can attempt to add one if you think it should have one.

### CODE REVIEW

**Comment by johnwiseheart** [comment]
Hmm, fails node 8 because globby requires 10+. I can swap to an older version of globby which supports 8.

**Comment by adidahiya** [comment]
@johnwiseheart Node 8 is officially deprecated this month, I don't think you need to support it anymore

**Comment by ericanderson** [comment]
Agreed. We should drop support


Eric L Anderson
Sent from my iPhone

> On Jan 14, 2020, at 20:43, Adi Dahiya <notifications@github.com> wrote:
> 

﻿
@johnwiseheart Node 8 is officially deprecated this month, I don't think you need to support it anymore

File: packages/rules/src/nestedWorkspaces.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    const expandedWorkspacesGlobs = await globby([...workspacePackageJsons, "!**/node_modules/**"]);
+
+    // Check that sets are equal (we know the inverse is true already)
+    const setsAreEqual = packageJsonPaths.every(packageJsonPath => expandedWorkspacesGlobs.includes(packageJsonPath));
+
+    if (!setsAreEqual) {
+      context.addError({
+        file: context.getPackageJsonPath(),
+        message:
+          'The "workspace" field is missing one or more values. This is usually because there is an unlisted nested workspace.' +
```
Comment (johnwiseheart) [inline packages/rules/src/nestedWorkspaces.ts]: include which ones are missing

File: packages/rules/src/nestedWorkspaces.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+import { Context, RuleModule } from "@monorepolint/core";
+import globby from "globby";
+import * as r from "runtypes";
+
+export const Options = r.Undefined;
+
+type Options = r.Static<typeof Options>;
+
+// Custom monorepolint rule to enforce that the root package.json contains
```
Comment (styu) [inline packages/rules/src/nestedWorkspaces.ts]:
```suggestion
// Enforces that the root package.json contains
```

File: packages/rules/src/nestedWorkspaces.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+export const nestedWorkspaces: RuleModule<typeof Options> = {
+  check: async (context: Context) => {
+    const rootPackageJson = context.getWorkspaceContext().getPackageJson();
+
+    // Expand a set of globs covering all package.json files in the entire repo (except the root)
+    const packageJsonPaths = await globby(["*/**/package.json", "!**/node_modules/**"]);
+
+    const workspaces = Array.isArray(rootPackageJson.workspaces)
+      ? rootPackageJson.workspaces
+      : rootPackageJson.workspaces !== undefined
```
Comment (styu) [inline packages/rules/src/nestedWorkspaces.ts]: oh sad, monorepolint is behind on typescript

Comment (johnwiseheart) [reply packages/rules/src/nestedWorkspaces.ts]: Yeah I wrote it so much nicer with optional chaining, but then was overcome by sadness when I realised it was on 3.5 :(

**Pull request merged**